### PR TITLE
Scatter Plot w Fit Lines Vignette

### PR DESCRIPTION
**Pull request by nkorinek**
Vignette to demonstrate line testing capabilities in the `base` package in `matplotcheck`. 

This vignette seems a bit sparse. I left out `assert_xydata` as it was demonstrrated in the `plot_testing_basics` vignette, but I could add it in here as a reminder if we wanted to! Also: I could add in x/y lims assertions in this vignette if we wanted to. I wasn't sure how focused we wanted this vignette to be, so I stuck to the basics and I have tests ready to add if we decide we want more!

### CODE REVIEW

**Comment by nkorinek** [comment]
@lwasser so this vignette is failing due to my use of seaborn to plot the regression line. Seaborn is not part of the CI environment. Do you want me to figure out a way to plot this without seaborn, or would we be able to add it to the environment? I know we've talked about not adding anymore packages to this environment, but thought I'd see what your opinion was. Thanks!

**Comment by lwasser** [comment]
@nkorinek ahhhhh we can add seaborn to our docs envt. i'm not sure if you can do it here.

https://github.com/earthlab/matplotcheck/blob/master/readthedocs.yml

the easiest would be to use something else. is that an option tho? some tools like geopandas do have a yml file just for read the docs.

**Comment by lwasser** [comment]
@nkorinek let's keep this simple. let's add seaborn as a dev-requirements item in the list of tools

**Comment by jlpalomino** [comment]
For our documentation, we have verbally discussed changes that will be completed by nkorinek, including the addition of more explanation about the object type that is being tested (e.g. a story about what kind of data is being used and what the test is actually looking for) both at the top of the notebook and in the comments, and a new section at the bottom of the vignette for additional options when using a Jupyter Notebook implementation.

**Comment by nkorinek** [comment]
@jlpalomino I'm hunting down a few issues on this branch, I'll ping you when they're ironed out

**Comment by nkorinek** [comment]
@jlpalomino I fixed the bugs I was seeing. CI is failing due to an unrelated issue and this vignette could use a review now!

File: CHANGELOG.md [hunk 1]
```diff
@@ -6,6 +6,8 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
 ## [Unreleased]
+* Added a vignette for line functions (@nkorinek, #155)
```
Comment (jlpalomino) [inline CHANGELOG.md]: Minor comment here - the style for these entries is present tense, so "Add" and "Change/Modify/Update".

File: matplotcheck/base.py [hunk 2]
```diff
@@ -1102,7 +1102,13 @@ def assert_line(
             ):
                 flag_exist = True
                 line_x_vals = [coord[0] for coord in path_verts]
-                if min(line_x_vals) <= min_val and max(line_x_vals) >= max_val:
+                if (
```
Comment (jlpalomino) [inline matplotcheck/base.py]: I assume this change in base.py is something that you and @lwasser discussed.  I will ask her to review this section to verify the changes.

File: examples/plot_line_testing.py [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+==================
+
+These are some examples of using the basic functionality of MatPlotCheck
+to test line plots in Python.
+"""
+
+################################################################################
+# Setup
+# -----
+# You will start by importing the required packages and plotting a creating
```
Comment (jlpalomino) [inline examples/plot_line_testing.py]: Use either "plotting" or "creating" a line plot.

File: examples/plot_line_testing.py [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+# a line plot.
+
+import numpy as np
+from scipy import stats
+import seaborn as sns
+import pandas as pd
+import matplotlib.pyplot as plt
```
Comment (jlpalomino) [inline examples/plot_line_testing.py]: Update the order for these imports, per PEP 8: matplotlib, seaborn, numpy, scipy, pandas, matplotcheck

File: examples/plot_line_testing.py [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+import matplotcheck.base as pt
+
+################################################################################
+# Create Example Data
+# -------------------
+# Before we can grade a plot we have to create example data. This data could be
```
Comment (jlpalomino) [inline examples/plot_line_testing.py]: Change to: "Before we create a plot, we need to create some data.  For plots with regression lines, you will need data in which you are looking for trends over time, such as maximum values from lidar-derived measurements."

File: examples/plot_line_testing.py [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+# maximum lidar height readings in an area over time, or some other data
+# in which you are looking for a trend over time.
+
+# Create the data for the line plot
+
```
Comment (jlpalomino) [inline examples/plot_line_testing.py]: Remove blank line after regular comment (i.e. code directly follows comment).

File: examples/plot_line_testing.py [hunk 7]
```diff
@@ -0,0 +1,137 @@
+"""
+Testing Line Plots
+==================
+
+These are some examples of using the basic functionality of MatPlotCheck
+to test line plots in Python.
```
Comment (jlpalomino) [inline examples/plot_line_testing.py]: Change to: "to test line plots (including regression lines) in Python."

File: examples/plot_line_testing.py [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+import matplotcheck.base as pt
+
+################################################################################
+# Create Example Data
+# -------------------
+# Before we can grade a plot we have to create example data. This data could be
+# maximum lidar height readings in an area over time, or some other data
+# in which you are looking for a trend over time.
+
+# Create the data for the line plot
```
Comment (jlpalomino) [inline examples/plot_line_testing.py]: "Create dataframe of data points"

File: examples/plot_line_testing.py [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+
+col1 = list(np.random.randint(25, size=15))
+col2 = list(np.random.randint(25, size=15))
+data = pd.DataFrame(list(zip(col1, col2)), columns=['Data1', 'Data2'])
+
+# Plot the points, line of regression, and a one to one line for reference
```
Comment (jlpalomino) [inline examples/plot_line_testing.py]: "Plot data points, regression line, and one-to-one (1:1) line for reference"

File: examples/plot_line_testing.py [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+
```
Comment (jlpalomino) [inline examples/plot_line_testing.py:32]: Remove blank line between comment and the next code line.

File: examples/plot_line_testing.py [hunk 11]
```diff
@@ -0,0 +1,137 @@
+"""
+Testing Line Plots
+==================
+
+These are some examples of using the basic functionality of MatPlotCheck
+to test line plots in Python.
+"""
+
+################################################################################
+# Setup
```
Comment (jlpalomino) [inline examples/plot_line_testing.py]: Change to: " # Import Packages"

File: examples/plot_line_testing.py [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+ax.set(xlabel='Data1',
+       ylabel='Data2',
+       title='Example Data Regression Plot',
+       xlim=(0, 25),
+       ylim=(0, 25))
+
+plt.show()
+
+# Convert axes object into a PlotTester object
+
```
Comment (jlpalomino) [inline examples/plot_line_testing.py:60]: Remove blank line between comment and the next code line.

File: examples/plot_line_testing.py [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+line_figure_tests = pt.PlotTester(ax)
+
+################################################################################
+# Testing the Line Plot
```
Comment (jlpalomino) [inline examples/plot_line_testing.py]: Move this section "Testing the Line Plot" to before line 56, so that the code to create the PlotTester object is within this text section.

File: examples/plot_line_testing.py [hunk 14]
```diff
@@ -0,0 +1,137 @@
...
+# ---------------------
+# Now you can make a PlotTester object and test the line plot. You can test
+# the line type is a one to one line or a regression line, and you can test
```
Comment (jlpalomino) [inline examples/plot_line_testing.py]: Change to: "that the line type is..."

File: examples/plot_line_testing.py [hunk 15]
```diff
@@ -0,0 +1,137 @@
...
+# that the line has the correct y intercept and slope.
+
+################################################################################
+# Testing the Line Types
+# ----------------------
+# As you can see, there are two line types on this plot. A one to one line for
```
Comment (jlpalomino) [inline examples/plot_line_testing.py]: Change to: "As you can see, there are two line types on this plot: a one-to-one line for reference and a regression line of the data points."

File: examples/plot_line_testing.py [hunk 16]
```diff
@@ -0,0 +1,137 @@
...
+# reference, and a regression line of the points plotted. You can use the
+# function ``assert_lines_of_type()`` to test if a one to one or regression line
+# (or both types of line) are present in your plot. These are the only types
```
Comment (jlpalomino) [inline examples/plot_line_testing.py]: Add a set of parenthesis around "These are the only...."

File: examples/plot_line_testing.py [hunk 17]
```diff
@@ -0,0 +1,137 @@
...
+
+col1 = list(np.random.randint(25, size=15))
+col2 = list(np.random.randint(25, size=15))
+data = pd.DataFrame(list(zip(col1, col2)), columns=['Data1', 'Data2'])
+
+# Plot the points, line of regression, and a one to one line for reference
+
+fig, ax = plt.subplots()
+
+# Points and line of regression
```
Comment (jlpalomino) [inline examples/plot_line_testing.py]: Change to: "and regression line"

File: examples/plot_line_testing.py [hunk 18]
```diff
@@ -0,0 +1,137 @@
...
+# (or both types of line) are present in your plot. These are the only types
+# of lines you can currently test for with this function.
+
+# Check line types
+line_figure_tests.assert_lines_of_type(line_types=['regression', 'onetoone'])
+
+################################################################################
+# Testing the Slope and Y Intercept
+# ---------------------------------
+# Another aspect of a line plot that you can test is the slope and Y intercept,
```
Comment (jlpalomino) [inline examples/plot_line_testing.py]: Change to: "Other aspects of the line plot that you can test are the slope and y intercept to ensure that the line represents the correct values".

File: examples/plot_line_testing.py [hunk 19]
```diff
@@ -0,0 +1,137 @@
...
+# which checks to ensure the line has the correct values. If you made your
+# line from a list of vertices, you can use the ``get_slope_yintercept()``
+# function in the PlotTester object in order to get the slope and y intercept.
```
Comment (jlpalomino) [inline examples/plot_line_testing.py]: Is the code called like this: ``line_figure_tests.get_slope_yintercept()``?  If so, I would specify that object name as an example, and also change the language of line 85 slightly to be: "method of the PlotTester object, in order to..."

File: examples/plot_line_testing.py [hunk 20]
```diff
@@ -0,0 +1,137 @@
...
+# However, if you made your line from a regression function, it will take an
+# extra step to get the slope and intercept data. In this example,
+# ``stats.linregress`` is used to calculate the slope and intercept data. Once
+# you have created that data, you can plug it into the ``assert_line()``
+# function to ensure your line in your plot has the correct values.
+
+# Create the slope and intercept data for the line in the plot to check against
```
Comment (jlpalomino) [inline examples/plot_line_testing.py]: Change to: "Get slope and y intercept data of regression line for testing"

File: examples/plot_line_testing.py [hunk 21]
```diff
@@ -0,0 +1,137 @@
...
+slope_data, intercept_data, _, _, _ = stats.linregress(
+    data.Data1, data.Data2)
+
+# Check line is correct
```
Comment (jlpalomino) [inline examples/plot_line_testing.py]: Change to: "Check that slope and y intercept are correct (expected) values"

File: examples/plot_line_testing.py [hunk 22]
```diff
@@ -0,0 +1,137 @@
...
+# 1:1 line
+ax.plot((0, 1), (0, 1), transform=ax.transAxes, ls='--', c='k')
+
+ax.set(xlabel='Data1',
+       ylabel='Data2',
+       title='Example Data Regression Plot',
+       xlim=(0, 25),
+       ylim=(0, 25));
+
+# HERE'S WHERE YOU STORE THE PLOT!
```
Comment (jlpalomino) [inline examples/plot_line_testing.py]: Change this text to match [vector vignette](https://github.com/nkorinek/matplotcheck/blob/da924a0e4374b1563120b517ae7a935fb07aa5a5/examples/plot_vector_testing.py): "Here is where you access the axes objects of the plot for testing.
You can add the code line below to the end of any plot cell to store all axes objects created by matplotlib in that cell."

File: examples/plot_line_testing.py [hunk 23]
```diff
@@ -0,0 +1,137 @@
...
+# This line at the end of a cell you are expecting a plot in will store any
+# matplotlib plot made in that cell so you can test it at a later time.
+plot_test_hold = nb.convert_axes(plt, which_axes="current")
+
+# This can then be turned into a PlotTester object.
```
Comment (jlpalomino) [inline examples/plot_line_testing.py]: Change this section to match [vector vignette](https://github.com/nkorinek/matplotcheck/blob/da924a0e4374b1563120b517ae7a935fb07aa5a5/examples/plot_vector_testing.py): "This object can then be turned into a PlotTester object."

File: examples/plot_line_testing.py [hunk 24]
```diff
@@ -0,0 +1,137 @@
...
+# Below is an example of how you could store the Axes of a plot you are hoping
+# to grade in a notebook.
+
+# First, import the Notebook module from MatPlotCheck
+import matplotcheck.notebook as nb
+
+# Plot the data
+fig, ax = plt.subplots()
+
+# Points and line of regression
```
Comment (jlpalomino) [inline examples/plot_line_testing.py]: Change to: "and regression line"

File: examples/plot_line_testing.py [hunk 25]
```diff
@@ -0,0 +1,137 @@
...
+    data.Data1, data.Data2)
+
+# Check line is correct
+line_figure_tests.assert_line(slope_exp=slope_data, intercept_exp=intercept_data)
+
+
+################################################################################
+# Access the Axes object in a Jupyter Notebook
+# --------------------------------------------
+# MatPlotCheck can be used to help grade Jupyter Notebooks as well. The main
```
Comment (jlpalomino) [inline examples/plot_line_testing.py]: Change this text to match [vector vignette](https://github.com/nkorinek/matplotcheck/blob/da924a0e4374b1563120b517ae7a935fb07aa5a5/examples/plot_vector_testing.py): "Matplotcheck can be used to test plots in Jupyter Notebooks as well. The main difference is how you access the axes objects from the plot that you want to test. Below is an example of how you could access the axes of a plot you want to test in a Jupyter Notebook."

**Review by jlpalomino — CHANGES_REQUESTED** [review]
Great job @nkorinek!  Mostly minor comments regarding wording, but let me know if you have any questions.  Ping me when you are ready for me to review again.  Thanks!

File: matplotcheck/base.py (hunk 2 above)
Comment (nkorinek) [reply matplotcheck/base.py]: @jlpalomino this is something that @lwasser and I have discussed! It's actually changes from another pull request https://github.com/earthlab/matplotcheck/pull/248 . I thought I had left a comment noting that this branch needs to wait to be merged until #248 is merged, but looking over my comments I may have forgotten to leave one, oops! All changes in base are from that PR.

**Comment by nkorinek** [comment]
@lwasser this vignette has been updated with the changes suggested by @jlpalomino

File: examples/plot_line_testing.py [hunk 26]
```diff
@@ -0,0 +1,134 @@
+"""
+Testing Line Plots
+==================
+
+These are some examples of using the basic functionality of MatPlotCheck
```
Comment (jlpalomino) [inline examples/plot_line_testing.py]: Minor change here that I missed before: Matplotcheck rather than MatPlotCheck.

File: examples/plot_line_testing.py [hunk 27]
```diff
@@ -0,0 +1,134 @@
...
+ax.set(xlabel='Data1',
+       ylabel='Data2',
+       title='Example Data Regression Plot',
+       xlim=(0, 25),
+       ylim=(0, 25))
+
+plt.show()
+
+################################################################################
+# Testing the Line Plot
```
Comment (jlpalomino) [inline examples/plot_line_testing.py]: Minor change here: "Test Line Plots", rather than "Testing the Line Plot"

File: examples/plot_line_testing.py [hunk 28]
```diff
@@ -0,0 +1,134 @@
...
+
+################################################################################
+# Access the Axes object in a Jupyter Notebook
+# --------------------------------------------
+# Matplotcheck can be used to test plots in Jupyter Notebooks as well. The main
+# difference is how you access the axes objects from the plot that you want to
+# test. Below is an example of how you could access the axes of a plot you want
+# to test in a Jupyter Notebook.
+
+# First, import the Notebook module from MatPlotCheck
```
Comment (jlpalomino) [inline examples/plot_line_testing.py]: Minor change here that I missed before: Matplotcheck rather than MatPlotCheck.

**Review by jlpalomino — CHANGES_REQUESTED** [review]
Thanks @nkorinek A few super small changes, and then I will approve the PR for Leah to do a final review.

**Comment by nkorinek** [comment]
@jlpalomino fixes are in!

**Review by jlpalomino — APPROVED** [review]
Great job @nkorinek! @lwasser This PR is ready for your review.

File: examples/plot_line_testing.py [hunk 29]
```diff
@@ -0,0 +1,134 @@
+"""
+Testing Line Plots
```
Comment (jlpalomino) [inline examples/plot_line_testing.py]: Oh goodness, missed one. Sorry @nkorinek  This needs a more complete title: "Test Line Plots with Matplotcheck"

**Comment by nkorinek** [comment]
@jlpalomino title changed!

File: examples/plot_line_testing.py [hunk 30]
```diff
@@ -0,0 +1,134 @@
...
+ax.set(xlabel='Data1',
+       ylabel='Data2',
+       title='Example Data Regression Plot',
+       xlim=(0, 25),
+       ylim=(0, 25))
+
+plt.show()
+
+################################################################################
+# Test Line Plots
```
Comment (lwasser) [inline examples/plot_line_testing.py]:
```suggestion
# Test Line Plots Using a basic matplotcheck.PlotTester object
```

File: examples/plot_line_testing.py [hunk 31]
```diff
@@ -0,0 +1,134 @@
...
+# Convert matplotlib plot axes object into a matplotcheck PlotTester object
+line_figure_tests = pt.PlotTester(ax)
+
+################################################################################
+# Test Line Types
+# ---------------
+# There are two line types on the plot above: a one-to-one line for
+# reference and a regression line derived from the data points. You can use the method
+# ``assert_lines_of_type()`` to test if a one to one or regression line
+# (or both line types) are present in the plot. (NOTE: Regression and 1:1 lines are the only types
```
Comment (lwasser) [inline examples/plot_line_testing.py]: @nkorinek i'm confused by this. In theory a line is a line. what does the regression line test actually do that is different from testing any other straight line? i'm just confused by the statement - you can test 1:1 and regression. both are lines so both have slopes and intercepts. are they stored differently in the MPL object? let's just chat a bit so i better understand! can MPC test for any line if there is a slope and intercept given?

File: examples/plot_line_testing.py [hunk 32]
```diff
@@ -0,0 +1,134 @@
...
+# Import Packages
+# ---------------
+# You will start by importing the required packages and plotting a line plot.
+
+import matplotlib.pyplot as plt
+import seaborn as sns
+import numpy as np
+from scipy import stats
+import pandas as pd
+import matplotcheck.base as pt
```
Comment (lwasser) [inline examples/plot_line_testing.py:18]: This is not an ideal import convention `pt`. 

Nathan - let's look into being able to import ANY of the mpc tester objects using
`mpc.RasterTester`, `mpc.VectorTester`, `mpc.PlotTester`

i think we can do this by modifying the init file or similar to what they do in geopandas??

File: examples/plot_line_testing.py [hunk 33]
```diff
@@ -0,0 +1,134 @@
...
+# Create Example Data
+# -------------------
+# Before you create a plot, you need to create some data. For plots with
+# regression lines, you will need data in which you are looking for trends
+# over time, such as maximum values from lidar-derived measurements.
+
+# Create dataframe of data points
+col1 = list(np.random.randint(25, size=15))
+col2 = list(np.random.randint(25, size=15))
+data = pd.DataFrame(list(zip(col1, col2)), columns=['Data1', 'Data2'])
```
Comment (lwasser) [inline examples/plot_line_testing.py]: minor but let's keep all column names lower case so data1 and data2 vs Data1
i like to be consistent with case in code even if it's column names!

File: examples/plot_line_testing.py [hunk 34]
```diff
@@ -0,0 +1,134 @@
...
+
+# Get slope and y intercept data of regression line for testing
+slope_data, intercept_data, _, _, _ = stats.linregress(
+    data.Data1, data.Data2)
+
+# Check that slope and y intercept are correct (expected) values
+line_figure_tests.assert_line(slope_exp=slope_data, intercept_exp=intercept_data)
+
+
+################################################################################
```
Comment (lwasser) [inline examples/plot_line_testing.py:102]: I will look at this when we fix the other imports in a locally build but i'm wondering if this should be a NOTE rather then a section at the bottom of the page. just so it stands out.  what do you think @nkorinek ?

File: matplotcheck/base.py (hunk 2 above)
Comment (lwasser) [reply matplotcheck/base.py]: good catch @jlpalomino  :) @nkorinek is making so many changes to this package that there are a lot of outstanding pr's that i need to look at :) and some need to be merged in specific order!

**Review by lwasser — CHANGES_REQUESTED** [review]
ok @nkorinek i had some questions here related to how we access MPC objects. 
have a look here at Jess's code. i think that we can do something like this to make imports cleaner! 
https://github.com/jhamrick/plotchecker/blob/master/plotchecker/__init__.py

now i know this will result in a need to update our homework tests and vignettes but i think it's worth it. i know there are a few open PR's so let's talk about what order we want to implement this! i'm open to your cleaning this up and then merging it and opening a new issue about imports. or i'm open to implementing this here but we will then need another pr to fix other vignettes. my brain is dead now so i can't decide what is best but you tell me how you'd like to implement this all!! (in what order)

File: examples/plot_line_testing.py (hunk 31 above)
Comment (nkorinek) [reply examples/plot_line_testing.py]: Hey @lwasser! This section became a sort of word soup, so sorry for any confusion. What I'm saying here is that the function `assert_lines_of_type()` can only check if a line is stored as a regression or a 1:1 line. I got that from the docstring describing the `line_types` argument, here: 
```
line_types : list of strings
            Acceptable strings in line_types are as follows
            ``['regression', 'onetoone']``.
```
The code than checks that there's a line which matches the slope/intercept of the line of regression of all points plotted, and/or that there's a line with a slope of 1 and an intercept of 0. That's what I was trying to point out in this section, but I could likely reword it to be more clear.

File: examples/plot_line_testing.py (hunk 34 above)
Comment (nkorinek) [reply examples/plot_line_testing.py:102]: @lwasser this could definitely be a note

**Review by nkorinek — COMMENTED** [review]
@lwasser Forgot to leave this comment earlier, sorry! I think it would be best if I opened up a new PR for the import changes and modified this along with the rest of the code in that PR. It makes more sense to me. Let me know if that works!

**Comment by nkorinek** [comment]
Update: this now relies on #274 to be merged first

**Review by lwasser — CHANGES_REQUESTED** [review]
ok @nkorinek a few more things on this vignette . When i build it i see this in the output. i'm just wondering if we use plt.show() in other vignettes as this seems to cause an issue given we have agg settings to ensure CI doesn't get confused. I suspect we don't want to use `plt.show()` in this vignette.

This vignette is making more sense to me. it's not actually a line plot vignette. it's a way to test lines on scatter plots. Given that i think we need a section above the jupyter section that shows the user how to 

1. test the point data on the plot
2. test the x and y axis labels
3. test the plot title.

This can be a fairly simple section but it would be best if people had everything they need to test a scatter plot with those two lines on it in one place. we then may rename the vignette. Take a stab at this and i'll look at it again. 

Screen Shot 2020-05-13 at 5 04 05 PM

**Comment by nkorinek** [comment]
@lwasser updated this vignette as per our discussion earlier today. Two important things to note:

1. I ran into a bug with `assert_xydata()` that I've never seen before. It was failing two identical numpy arrays saying that there shapes didn't match. One has a shape of `(15, 1)` while the other had a shape of `(15,)`. So they were the same array, it's just that one was nested inside the other. I fixed this by adding `.flatten()` to the end of the checks to make all of the data checks one dimensional. This will still fail data that doesn't match in length, but fixes this issue. Hope that's ok! Just a note: these arrays should never be more than one dimensional since it's checking the individual columns of a dataframe.

2. I know we discussed `check-coverage` a lot, however, I found an instance where I think the check should be automatic. Not to say we should take the argument out of the users hands, but when you want to check a regression line and a onetoone line at the same time, it causes issues. You can't set the `check_coverage` to false, as it won't check the coverage of the regression line. However, if you leave it as true than the onetoone line will make the test fail. SO. I think it should be set to `False` automatically if the line is a onetoone, but we let users still decide if they want to check coverage on the regression line. Let me know if that makes sense!
